tests/tests_pytorch/strategies/test_xla_fsdp.py

```diff
@@ -0,0 +1,183 @@
+# Copyright The Lightning AI team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import re
+from typing import Optional
+from unittest import mock
+from unittest.mock import Mock
+
+import pytest
+import torch
+from lightning.fabric.strategies.xla_fsdp import _activation_checkpointing_auto_wrapper
+from lightning.pytorch import Trainer
+from lightning.pytorch.demos.boring_classes import BoringModel
+from lightning.pytorch.plugins.precision import XLAPrecisionPlugin
+from lightning.pytorch.strategies import XLAFSDPStrategy
+
+from tests_pytorch.helpers.runif import RunIf
+
+
+class TestFSDPModel(BoringModel):
+    def __init__(self):
+        super().__init__()
+        self.layer: Optional[torch.nn.Module] = None
+
+    def _init_model(self) -> None:
+        self.layer = torch.nn.Sequential(torch.nn.Linear(32, 32), torch.nn.ReLU(), torch.nn.Linear(32, 2))
+
+    def configure_model(self) -> None:
+        if self.layer is None:
+            self._init_model()
+        # the model is already wrapped with FSDP: no need to wrap again!
+        from torch_xla.distributed.fsdp.xla_fully_sharded_data_parallel import XlaFullyShardedDataParallel
+
+        if isinstance(self.layer, XlaFullyShardedDataParallel):
+            return
+        for i, layer in enumerate(self.layer):
+            if i % 2 == 0:
+                self.layer[i] = XlaFullyShardedDataParallel(layer)
+        self.layer = XlaFullyShardedDataParallel(self.layer)
+
+    def configure_optimizers(self):
+        return torch.optim.AdamW(self.layer.parameters(), lr=0.1)
+
+    def on_train_batch_start(self, batch, batch_idx):
+        assert batch.dtype == torch.float32
+
+
+def _run_multiple_stages(trainer, model, state_dict_type, model_path: Optional[str] = None):
+    local_process_count = len(trainer.strategy.parallel_devices)
+    world_size = 4
```

Comment:
@carmocca FYI, when I used the strategy's `world_size` definition this test would freeze, so I ended up hardcoding this.

Reply:
This is expected because at this point, we aren't in the launched region so `world_size` will return 1 and that produces the deadlock.